parser/parser.go

```diff
@@ -149,17 +149,27 @@ func Pin(ctx context.Context, res resolver.Resolver, parser Parser, nodes []*yam
 				return
 			}
 
-			resolved, err := res.Resolve(ctx, ref)
+			newRef := ref
+			if upgrade {
+				if newRef, err = res.Upgrade(ctx, newRef); err != nil {
```

Comment:
I wonder if `upgrade` should return the resolved value too? That would save us an additional lookup.